### ISSUE TITLE
In javascript front end, websocket updates to builds don't include properties requested

### ISSUE THREAD

**Issue by uglycoyote** [buildbot/buildbot#4018]
I have noticed that when we subscribe to updates of builds via code like you find in ```console_view``` (taken from [this file](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee))

        @$scope.builds = @builds = @dataAccessor.getBuilds
            property: ["got_revision"]
            limit: @buildLimit
            order: '-started_at'

the initial response will include the requested extra properties (```got_revision``` in this case) but subsequent updates to the set of builds that come over via websocket will have an empty dictionary of properties and will not get the ```got_revision```.

You can see this in Chrome's devtools, under the network tab while viewing the Console page:

- look at the response to the ```builds``` request, if you search the response you will find ```got_revision``` properties
- look at the ```ws```, under **Frames**,  and wait until a frame appears with ```k: "builds/NNNN/new"```, and observe that the build's data comes through with ```properties: {}```

While developing our custom version of the ```console_view``` plugin, I attempted to simplify the code by getting rid of the queries for ```buildsets``` and ```buildrequests``` and relying on the ```got_revision```  property.  If the ```got_revision``` property comes through, that is sufficient to match up the builds with changes.  But this failed to work out, because of new build data coming over the websocket without ```got_revision``` in it.  

On my custom plugin I would also like to display (at the top level) additional information about the builds which is also contained in the build's properties.  Without going into specifics, our custom console would have icons within the "badges" for each build, where the icon image would depend on the value of some property which is already present on the build -- a custom value that we insert during the build process (i.e., I can see the property in builders/NN/builds/NNNN under the **Build Properties** tab)

Ideally I would just be able to add my desired property to the ```property``` list in the snippet above, passed in to ```dataAccessor.getBuilds```.  However, due to this problem with the websocket updates, that property will stop coming after the initial response.   

Due to this restriction (bug?) the only way I could implement my icon feature would be to make individual requests to get the details of each build that comes in, which seems like an unreasonable amount of traffic.

**Comment by tardyp** [comment]
You are right, the ws protocol does not allow data enhancement like the api does.

So dashboards that use property cannot be really live unless they force load properties upon new events.  We can either:

- change the ws API to allow automatic enhancement of events (i.e loading properties)
- change the data-module (javascript API) in order to automatically load properties upon events

The latter is probably the easiest, but as you said it comes with the price of lots of http requests

**Comment by uglycoyote** [comment]
This issue continues to vex us in various ways.   If we wanted to investigate enhancing the websocket API to allow additional properties to be sent with the updates, can you give us some pointers as to where that might happen in the code?

**Comment by uglycoyote** [comment]
From a peek at the code, it looks like the `WsProtocol` in `ws.py` has a `cmd_startconsuming` which calls  `self.master.mq.startConsuming`, which I assume is tied in to the `WampMQ.startConsuming` implementation?

So I assume it would come down to having the things which call `self.master.mq.produce` add the extra properties to events, such as builds.py

**Comment by tardyp** [comment]
graphql api already supports fetching additional properties in the event mecanism.
Adding properties to build event would consume a lot of network and memory bandwith.

There is a build event for every build status update. putting properties inside it will increase the size of event by one or two order of magnitude.

**Comment by uglycoyote** [comment]
> putting properties inside it will increase the size of event by one or two order of magnitude.

There's really only one property that we are usually interested in, which is the `revision` property.  Adding that property to each build record sent over websocket would increase the size of the record form ~275 bytes to ~300, bytes or a 10% increase in traffic on those requests.    Furthermore we don't really need it on every build status update, only on the first time where a build is introduced.  

Currently due to the absence of this information, we need to subscribe via websockets to the `buildsets` and `buildrequests` solely for the purpose of connecting builds with revisions. 

Based on a short observational period, I collected all of the websocket records in a variable `m` and filtered them to determine how much bandwidth is being taken up by each type of record

```
m.filter(x=>x.type=="receive").filter(x=>x.data.indexOf("buildrequests/")!=-1).map(x=>x.data.toString().length).reduce((x,y)=>x+y,0)
28875
m.filter(x=>x.type=="receive").filter(x=>x.data.indexOf("builds/")!=-1).map(x=>x.data.toString().length).reduce((x,y)=>x+y,0)
45030
m.filter(x=>x.type=="receive").filter(x=>x.data.indexOf("buildsets/")!=-1).map(x=>x.data.toString().length).reduce((x,y)=>x+y,0)
32791
```

So, we would reduce our websocket traffic to less than half if we were able to get the `revision` property to come through with the builds and not need to request `buildrequests` and `buildsets`

Anyhow, to us the bandwidth issue is a non-issue... The bigger issue is maintaining two different code paths in the front end code, one where the build objects have the `revision` property (because they are coming via a GET request) and one where the build objects don't have the `revision` property (because they are coming via websocket updates).

**Comment by tardyp** [comment]
> There's really only one property that we are usually interested in, which is the revision property

That is indeed the usual usecase, a typical web view only requires a couple of properties.
but the build event is generated deep inside the buildbot core.
We cannot just insert the revision property. Other views may require owner property, etc.

This is why graphql is so useful, you can just request the property that you need.

There is no test for this in https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/integration/test_graphql_queries.yaml

I thought I verified this important usecase was working...

**Comment by uglycoyote** [comment]
>  We cannot just insert the revision property. Other views may require owner property, etc.

When requesting properties over the REST API, the view must explicitly request the properties that it wants, e.g

```api/v2/builds?limit=5000&order=-started_at&property=revision```

What I would propose is that it should be possible to to subscribe to websocket events in the same way, where each subscriber subscribes to the information they need, and the server knows what scope of information to send to the subscribers.   Consumers should not need to receive information that they aren't interested in.    

Currently we have the opposite situation, where consumer's can't get any of the properties they want via websocket.  That's great for bandwidth... only until they need to go and shoot off a whole bunch of extra queries to get those properties separately.

> This is why graphql is so useful, you can just request the property that you need.

You have mentioned GraphQL a couple times, but to be honest I'm not sure how that fits in to the picture here.  I searched for GraphQL in the buildbot documentation but only found it mentioned once in relation to Github authentication (whether the auth client should use use Github's REST vs GraphQL API to authenticate).

Does Buildbot have a GraphQL server which is the newer fancier replacement for the REST API?

For context, we have a custom javascript/React console view which is based on `console_view`, and so does a similar kind of usage of the existing `dataAccessor`, which does an initial REST query to get the information up front, and then listens to websockets to get the `onChange`'s

```
        this.changes = this.dataAccessor.getChanges({limit: this.changeLimit, order: '-changeid'});
        this.buildrequests = this.dataAccessor.getBuildrequests({limit: this.buildLimit, order: '-submitted_at'});
        this.buildsets = this.dataAccessor.getBuildsets({limit: this.buildLimit, order: '-submitted_at'});

        this.builds.onChange = this.onChange;
        this.changes.onChange = this.onChange;
        this.buildrequests.onChange = this.onChange;
        this.buildsets.onChange = this.onChange;
```

In the context of a web front end, Is you recommendation to replace this with a Javascript GraphQL client like [Apollo](https://www.apollographql.com/docs/react/), or [Relay](https://relay.dev/)?

As I understand it, those libraries are still using websockets receiving updates to events that are subscribed to.  Would moving to GraphQL solve this issue of being able to subscribe to a specific set of properties on objects like a `build` and continuiously receive them via the websocket updates?

Are there any samples in the buildbot repo of GraphQL queries done from the web front-end via Javascript?  

Thanks!

**Comment by tardyp** [comment]
I have been indeed working on graphql support as an api v3 for a while.
I linked on a few graphql example queries here:
https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/integration/test_graphql_queries.yaml

The server side is mostly done, but indeed the client side has to be written. I've written some todo tasks here:
https://github.com/buildbot/buildbot/projects/5
The event support is there already and allow arbitrary query to be subscribed to.

I have made some tests with gqless which has a very interesting programming model, but maybe we should rather stick with apollo, which is probably much more mature.

This is part of a larger plan to transition out of angularjs which is deprecated.

**Comment by tardyp** [comment]
With this topic, I realized I did not push my property over graphql changes yet.
I finished it today. You can see some example of how you can subscribe to specific properties here:
https://github.com/buildbot/buildbot/pull/6260/files#diff-78e75d2fa2732f4eeb4a9ddf9ae2095446bfbba6732582ffdd0f1ff691b33417R155

about subscription, indeed, graphql has subscription already defined, which avoids us to invent another one.
In buildbot, you will be able to subscribe to whatever you can query. buildbot will maintain a list of those subscription and recalculate it on demand. 
This is what hasura claims to do already, I am not sure if it will be scalable, but with proper debouncing I think this can work and this will make the programming model very simple to use, no need to figure out what you can subscribe to and what you cannot.

**Comment by uglycoyote** [comment]
Thanks!  that looks promising!
